src/big/big_int.cr

```diff
@@ -410,36 +410,92 @@ struct BigInt < Int
+      "1"
+    else
+      count = LibGMP.sizeinbase(self, base).to_i
+      negative = self < 0
+
+      if precision <= count
+        len = count + (negative ? 1 : 0)
+        String.new(len + 1) do |buffer| # null terminator required by GMP
+          LibGMP.get_str(buffer, upcase ? -base : base, self)
+          base62_swapcase(Slice.new(buffer, len)) if base == 62
+          {len, len}
+        end
+      else
+        len = precision + (negative ? 1 : 0)
+        String.new(len + 1) do |buffer|
+          # e.g. precision = 13, count = 8
+          # "_____12345678\0" for positive
+          # "_____-12345678\0" for negative
+          LibGMP.get_str(buffer + precision - count, upcase ? -base : base, self)
+
+          if negative
+            buffer.value = '-'.ord.to_u8
+            Intrinsics.memset(buffer + 1, '0'.ord.to_u8, precision - count, false)
+          else
+            Intrinsics.memset(buffer, '0'.ord.to_u8, precision - count, false)
+          end
+
+          base62_swapcase(Slice.new(buffer, len)) if base == 62
+          {len, len}
+        end
+      end
+    end
   end
 
-  # :ditto:
-  def to_s(io : IO) : Nil
-    str = to_cstr
-    io.write_utf8 Slice.new(str, LibC.strlen(str))
+  def to_s(io : IO, base : Int = 10, *, precision : Int = 1, upcase : Bool = false) : Nil
+    raise ArgumentError.new("Invalid base #{base}") unless 2 <= base <= 36 || base == 62
+    raise ArgumentError.new("upcase must be false for base 62") if upcase && base == 62
+    raise ArgumentError.new("Precision must be non-negative") unless precision >= 0
+
+    case {self, precision}
+    when {0, 0}
+      # do nothing
+    when {0, 1}
+      io << '0'
+    when {1, 1}
+      io << '1'
+    else
+      count = LibGMP.sizeinbase(self, base).to_i
+      ptr = LibGMP.get_str(nil, upcase ? -base : base, self)
+      negative = self < 0
+
+      if precision <= count
+        buffer = Slice.new(ptr, count + (negative ? 1 : 0))
+      elsif negative
+        io << '-'
+        (precision - count).times { io << '0' }
+        buffer = Slice.new(ptr + 1, count) # this is the absolute value
+      else
+        (precision - count).times { io << '0' }
+        buffer = Slice.new(ptr, count)
+      end
+
+      base62_swapcase(buffer) if base == 62
+      io.write_utf8 buffer
+    end
   end
 
-  # Returns a string containing the representation of big radix base (2 through 36).
-  #
-  # ```
-  # require "big"
-  #
-  # BigInt.new("123456789101101987654321").to_s(8)  # => "32111154373025463465765261"
-  # BigInt.new("123456789101101987654321").to_s(16) # => "1a249b1f61599cd7eab1"
-  # BigInt.new("123456789101101987654321").to_s(36) # => "k3qmt029k48nmpd"
-  # ```
-  def to_s(base : Int) : String
-    raise ArgumentError.new("Invalid base #{base}") unless 2 <= base <= 36
-    cstr = LibGMP.get_str(nil, base, self)
-    String.new(cstr)
+  private def base62_swapcase(buffer)
+    buffer.map! do |x|
+      if x & 0x40 != 0 # A..Z, a..z
+        x ^ 0x20
```

Comment:
Maybe add a comment that thes simplified condition and transformation are sufficient because `buffer` is guaranteed to contain only alphanumerical ASCII characters?